src/connectionController.ts

```diff
@@ -13,7 +13,7 @@ import ConnectionString from 'mongodb-connection-string-url';
 import { EventEmitter } from 'events';
 import type { MongoClientOptions } from 'mongodb';
 import { v4 as uuidv4 } from 'uuid';
-
+import { createKeytar } from './utils/keytar';
```

Comment:
Should we maybe update this model, so it remembers that keytar require failed once, and we don't try to resolve it again for each next connection that goes through the migration? With our webpack configuration, the failed require won't be cached: https://github.com/mongodb-js/vscode/blob/711587b9755127056d70b0ff91a482cb318957f1/webpack.config.js#L75-L76

Reply:
Another way to handle it is to move the createKeytar call to the loadSavedConnections function, where can see if there are connections without secretStorageLocation and then try to require it once for the whole migration attempt.

Reply:
I would require it once in loadSavedConnections, but you can decide what works better implementation-wise.

Reply:
Moving it to loadSavedConnections needs some changes in the tests, as some of them fail. 

I'll keep the current version and merge.

Thanks a lot!

Reply:
What will happen when vscode removes keytar from vscode and there are migrations that require a migration? Are we ok with resolving the keytar module for each connection? Will this make the launch of the extension slower than it could be?

Reply:
Moving createKeytar to loadSavedConnections should note brake tests. Have you tried something like this?

```typescript
const hasConnectionsThatDidNotMigrateEarlier =
      !!globalAndWorkspaceConnections.some(([, connectionInfo]) => !connectionInfo.storageLocation);

if (hasConnectionsThatDidNotMigrateEarlier) {
  try {
    ext.keytarModule =
    ext.keytarModule === undefined ? createKeytar() : ext.keytarModule;
  } catch (err) {
    // Couldn't load keytar, proceed without storing & loading connections.
  }
}
```